### PR TITLE
(GH-1204) Add option to stop the puppet agent service after install

### PR DESCRIPTION
**Pull request by lucywyman**
This adds a parameter to the `puppet_agent::install` task that, when
true, will stop the puppet service after install. The parameter defaults
to false. The option is available for both unix and windows platforms

### CODE REVIEW

**Comment by lucywyman** [comment]
Closes @puppetlabs/bolt#1204

File: tasks/install_powershell.ps1 [hunk 1]
```diff
@@ -3,7 +3,8 @@ Param(
 	[String]$version,
   [String]$collection = 'puppet',
   [String]$windows_source = 'https://downloads.puppet.com',
-  [String]$install_options = 'REINSTALLMODE="amus"'
+  [String]$install_options = 'REINSTALLMODE="amus"',
+  [String]$stop_service = 'false'
```
Comment (donoghuc) [inline tasks/install_powershell.ps1]: Should this be `[bool]$stop_service = $false`?

File: task_spec/spec/acceptance/nodesets/default.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+HOSTS:
+  centos-7-x64:
+    roles:
+      - target
+      - default
```
Comment (donoghuc) [inline task_spec/spec/acceptance/nodesets/default.yml]: Why was this added? I thought that travis tests used docker?

File: task_spec/spec/acceptance/init_spec.rb [hunk 3]
```diff
@@ -57,6 +59,10 @@ def target_platform
       expect(res['result']['source']).to be
     end
 
+    service = run_command('/opt/puppetlabs/bin/puppet resource service puppet-agent', 'target')
```
Comment (donoghuc) [inline task_spec/spec/acceptance/init_spec.rb]: I think we need a windows specific test as well.

**Comment by donoghuc** [comment]
Unfortunately the workflow for running the acceptance tests on a PR involve an experimental pipeline. I set that up with https://github.com/puppetlabs/ci-job-configs/pull/6306 , here is a link to the job https://jenkins-master-prod-1.delivery.puppetlabs.net/view/__experimental%20automatic/job/experimental_auto_lucywyman-puppet-agent-module_intn-run_task_acceptance-GH-1204-add-stop-service-option/

File: task_spec/spec/acceptance/init_spec.rb [hunk 4]
```diff
@@ -57,6 +59,15 @@ def target_platform
       expect(res['result']['source']).to be
     end
 
+    if target_platform =~ /win/
+      # & "Command" properly includes spaces
+      service = run_command('& "C:\Program Files\Puppet Labs\Puppet\bin\puppet.bat" resource service puppet-agent', 'target')
+    else
+      service = run_command('/opt/puppetlabs/bin/puppet resource service puppet-agent', 'target')
```
Comment (donoghuc) [inline task_spec/spec/acceptance/init_spec.rb]: I dont think we expect the macos install to start pxp-agent service.

File: tasks/install_shell.sh [hunk 5]
```diff
@@ -525,6 +525,10 @@ do_download "$download_url" "$download_filename"
 
 install_file $filetype "$download_filename"
 
+if [[ $PT_stop_service = true ]] && [ $platform != "mac_os_x" ]; then
+  /opt/puppetlabs/bin/puppet resource service puppet-agent ensure=stopped enable=false
```
Comment (donoghuc) [inline tasks/install_shell.sh]: I think it is OK to unconditionally stop the service. My comment above was more about updating the spec test to allow for `ensure => absent` in the service command result for osx (and window maybe I cant remember if it is started there).

**Comment by lucywyman** [comment]
@donoghuc  I'm a bit unclear on Windows as well. I installed it manually on an instance, and was able to successfully check the service status with `C:\Program Files\Puppet Labs\Puppet\bin\puppet.bat resource service puppet-agent`. However checking output in the tests I get:
```
[1] pry(#<RSpec::ExampleGroups::InstallTask>)> service
=> [{"node"=>"we3l5e002nw40o2.delivery.puppetlabs.net",
  "target"=>"we3l5e002nw40o2.delivery.puppetlabs.net",
  "action"=>"command",
  "object"=>"& \"C:\\Program Files\\Puppet Labs\\Puppet\\bin\\puppet.bat\" resource service puppet-agent",
  "status"=>"failure",
  "result"=>
   {"stdout"=>"",
    "stderr"=>
     "puppet.bat : C:/Program Files/Puppet Labs/Puppet/sys/ruby/lib/ruby/gems/2.4.0/gems/win32-service-0.8.8/lib/win32/service.rb:1081: warning: constant ::Fixnum is deprecated\n" +
     "    + CategoryInfo          : NotSpecified: (C:/Program File...m is deprecated:String) [], RemoteException\n" +
     "    + FullyQualifiedErrorId : NativeCommandError\n" +
     "WARNING: Failed to retrieve description for the WinQuic service.\r\n" +
     "WARNING: Unable to get delayed auto start information for the WinQuic service\r\n" +
     "WARNING: Unable to retrieve failure actions for the WinQuic service\r\n" +
     "\e[1;31mError: Could not run: Cannot get status of puppet-agent, error was: The specified service does not exist as an installed service. - OpenService: The specified service does not exist as an installed service.\r\n" +
     "Wrapped exception:\r\n" +
     "The specified service does not exist as an installed service. - OpenService: The specified service does not exist as an installed service.\e[0m\r\n",
    "exit_code"=>1,
    "_error"=>
     {"kind"=>"puppetlabs.tasks/command-error",
      "issue_code"=>"COMMAND_ERROR",
      "msg"=>"The command failed with exit code 1",
      "details"=>{"exit_code"=>1}}}}]
```
Which makes it seem like the service isn't even installed? I'm not sure if it's just not managed by Puppet though?

**Comment by donoghuc** [comment]
Did you try something like
```
run_command('c:/"program files"/"puppet labs"/puppet/bin/puppet resource service puppet')
```
I forget how the run_command in the boltspec run stuff will work, but the following plan seems to be fine:
```
plan test(TargetSpec $nodes){
  return run_command('c:/"program files"/"puppet labs"/puppet/bin/puppet resource service puppet', $nodes)
}
```
```
cas@cas-ThinkPad-T460p:~/working_dir/bolt$ bolt plan run test -t o4ffnatv68q9by4.delivery.puppetlabs.net --transport winrm --user Administrator --password Qu@lity! --no-ssl
Starting: plan test
Starting: plan test
Starting: command 'c:/"program files"/"puppet labs"/puppet/bin/puppet resource service puppet' on o4ffnatv68q9by4.delivery.puppetlabs.net
Finished: command 'c:/"program files"/"puppet labs"/puppet/bin/puppet resource service puppet' with 0 failures in 5.11 sec
Finished: plan test in 5.12 sec
Finished: plan test in 5.12 sec
Finished on o4ffnatv68q9by4.delivery.puppetlabs.net:
  STDOUT:
    service { 'puppet':
      ensure => 'running',
      enable => 'true',
    }
Successful on 1 node: o4ffnatv68q9by4.delivery.puppetlabs.net
Ran on 1 node
```
Also, why are you targetting the puppet-agent service? In apply_prep I thought we stop and disable `puppet` https://github.com/puppetlabs/bolt/blob/9df36dd1e55ccdc59d87e062711cb0992e17dcaa/lib/bolt/plugin/install_agent.rb#L20-L21

File: task_spec/spec/acceptance/init_spec.rb [hunk 6]
```diff
@@ -57,6 +59,15 @@ def target_platform
...
     end
 
+    service = if target_platform =~ /win/
+                run_command('c:/"program files"/"puppet labs"/puppet/bin/puppet resource service puppet', 'target')
+              else
+                run_command('/opt/puppetlabs/bin/puppet resource service puppet', 'target')
+              end
+    output = service[0]['result']['stdout']
+    expect(output).to include("ensure => 'stopped'")
+      .or include("ensure => 'absent'")
```
Comment (donoghuc) [inline task_spec/spec/acceptance/init_spec.rb]: Now that we are not stopping `pxp-agent` will `puppet` ever be `absent`?

Comment (lucywyman) [reply task_spec/spec/acceptance/init_spec.rb]: Seems like no!

**Comment by lucywyman** [comment]
https://jenkins-master-prod-1.delivery.puppetlabs.net/view/__experimental%20automatic/job/experimental_auto_lucywyman-puppet-agent-module_intn-run_task_acceptance-GH-1204-add-stop-service-option/ Latest run looks :green_heart:

**Comment by donoghuc** [comment]
Inventory:
```
version: 2
groups:
  - name: custom_nodes
    targets:
      - uri: wyxnk0rn4rewzu2.delivery.puppetlabs.net
    plugin_hooks:
      puppet_library:
        plugin: task
        task: puppet_agent::install
        parameters:
          stop_service: true
```
```
cas@cas-ThinkPad-T460p:~/working_dir/bolt$ bolt apply -e "notice('hi')" -t custom_nodes
Starting: plan 
Starting: install puppet and gather facts on wyxnk0rn4rewzu2.delivery.puppetlabs.net
Finished: install puppet and gather facts with 0 failures in 15.61 sec
Starting: apply catalog on wyxnk0rn4rewzu2.delivery.puppetlabs.net
wyxnk0rn4rewzu2.delivery.puppetlabs.net: Scope(Class[main]): hi
Started on wyxnk0rn4rewzu2.delivery.puppetlabs.net...
Finished on wyxnk0rn4rewzu2.delivery.puppetlabs.net:
  changed: 0, failed: 0, unchanged: 0 skipped: 0, noop: 0
Finished: apply catalog with 0 failures in 4.13 sec
Successful on 1 node: wyxnk0rn4rewzu2.delivery.puppetlabs.net
Ran on 1 node in 19.82 sec
cas@cas-ThinkPad-T460p:~/working_dir/bolt$ bolt task run service name=puppet action=status -t custom_nodes
Started on wyxnk0rn4rewzu2.delivery.puppetlabs.net...
Finished on wyxnk0rn4rewzu2.delivery.puppetlabs.net:
  {
    "status": "MainPID=0,LoadState=loaded,ActiveState=inactive",
    "enabled": "disabled"
  }
Successful on 1 node: wyxnk0rn4rewzu2.delivery.puppetlabs.net
Ran on 1 node in 0.67 sec
```
